scripts/bert/bert_qa_dataset.py

```diff
@@ -0,0 +1,516 @@
+
+            assert len(input_ids) == self.max_seq_length
+            assert len(segment_ids) == self.max_seq_length
+
+            start_position = 0
+            end_position = 0
+            if self.is_training and not example.is_impossible:
+                # For training, if our document chunk does not contain an annotation
+                # we throw it out, since there is nothing to predict.
+                doc_start = doc_span.start
+                doc_end = doc_span.start + doc_span.length - 1
+                out_of_span = False
+                if not (tok_start_position >= doc_start
+                        and tok_end_position <= doc_end):
+                    out_of_span = True
+                if out_of_span:
+                    start_position = 0
+                    end_position = 0
+                else:
+                    doc_offset = len(query_tokens) + 2
+                    start_position = tok_start_position - doc_start + doc_offset
+                    end_position = tok_end_position - doc_start + doc_offset
+
+            if self.is_training and example.is_impossible:
+                start_position = 0
+                end_position = 0
+            features.append(SquadFeature(example_id=example.example_id,
+                                         qas_id=example.qas_id,
+                                         doc_tokens=example.doc_tokens,
+                                         doc_span_index=doc_span_index,
+                                         tokens=tokens,
+                                         token_to_orig_map=token_to_orig_map,
+                                         token_is_max_context=token_is_max_context,
+                                         input_ids=input_ids,
+                                         valid_length=valid_length,
+                                         segment_ids=segment_ids,
+                                         start_position=start_position,
+                                         end_position=end_position,
+                                         is_impossible=example.is_impossible))
+        return features
+
+    def __call__(self, record):
+        examples = self._transform(*record)
+        if not examples:
+            return None
+        features = []
+
+        for _example in examples:
+            feature = []
+            feature.append(_example.example_id)
+            feature.append(_example.input_ids)
+            feature.append(_example.segment_ids)
+            feature.append(_example.valid_length)
+            feature.append(_example.start_position)
+            feature.append(_example.end_position)
+            features.append(feature)
+
+        return features
+
+
+def bert_qa_batchify_fn(data):
+    """Collate data into batch."""
+    def batchify_fn(data):
+        data = np.asarray(data)
+        return nd.array(data, dtype=data.dtype, ctx=context.Context('cpu_shared', 0))
```

Comment:
why are we using `ctx=context.Context('cpu_shared', 0)`?

Reply:
This is my reference to the implementation of gluon in mxnet